commit 72364659f16f415dec104112ee5fa286674d278f
Author: Jonathan Ellis <jbellis@apache.org>
Date:   Sun Feb 5 17:22:40 2012 -0600

    don't wait for migration_request reply, to avoid deadlock
    patch by Peter Schuller; reviewed by jbellis for CASSANDRA-3832

diff --git a/CHANGES.txt b/CHANGES.txt
index 48a8074..cf0b201 100644
--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -52,7 +52,7 @@
  * Allow rangeSlice queries to be start/end inclusive/exclusive (CASSANDRA-3749)
  * Fix BulkLoader to support new SSTable layout and add stream
    throttling to prevent an NPE when there is no yaml config (CASSANDRA-3752)
- * Allow concurrent schema migrations (CASSANDRA-1391)
+ * Allow concurrent schema migrations (CASSANDRA-1391, 3832)
  * Add SnapshotCommand to trigger snapshot on remote node (CASSANDRA-3721)
  * Make CFMetaData conversions to/from thrift/native schema inverses
    (CASSANDRA_3559)
diff --git a/src/java/org/apache/cassandra/service/MigrationManager.java b/src/java/org/apache/cassandra/service/MigrationManager.java
index 7771815..0d8df1d 100644
--- a/src/java/org/apache/cassandra/service/MigrationManager.java
+++ b/src/java/org/apache/cassandra/service/MigrationManager.java
@@ -91,10 +91,13 @@ public class MigrationManager implements IEndpointStateChangeSubscriber
 
         /**
          * if versions differ this node sends request with local migration list to the endpoint
-         * and expecting to receive a list of migrations to apply locally
+         * and expecting to receive a list of migrations to apply locally.
+         *
+         * Do not de-ref the future because that causes distributed deadlock (CASSANDRA-3832) because we are
+         * running in the gossip stage.
          */
 
-        Future f = StageManager.getStage(Stage.MIGRATION).submit(new WrappedRunnable()
+        StageManager.getStage(Stage.MIGRATION).submit(new WrappedRunnable()
         {
             public void runMayThrow() throws Exception
             {
@@ -128,8 +131,6 @@ public class MigrationManager implements IEndpointStateChangeSubscriber
                 }
             }
         });
-
-        FBUtilities.waitOnFuture(f);
     }
 
     public static boolean isReadyForBootstrap()
